src/HDF5Msg.chpl

```diff
@@ -896,6 +898,371 @@ module HDF5Msg {
+
+                        writeArkoudaMetaData(file_id, group, objType, getHDF5Type(uint));
+                    } when (DType.Float64) {
+                        var valEntry:SymEntry = toSymEntry(toGenSymEntry(vals), real);
+
+                        //localize values and write dataset
+                        var localVals = new lowLevelLocalizingSlice(valEntry.a, 0..#valEntry.size);
+                        var val_dims: C_HDF5.hsize_t = valEntry.size:C_HDF5.hsize_t;
+                        C_HDF5.H5LTmake_dataset(file_id, "/%s/values".format(group).c_str(), 1:c_int, val_dims, getHDF5Type(real), localVals.ptr);
+                        
+                        //localize segments and write dataset
+                        var localSegs = new lowLevelLocalizingSlice(segEntry.a, 0..#segEntry.size);
+                        var seg_dims: C_HDF5.hsize_t = segEntry.size:C_HDF5.hsize_t;
+                        C_HDF5.H5LTmake_dataset(file_id, "/%s/segments".format(group).c_str(), 1:c_int, seg_dims, getHDF5Type(int), localSegs.ptr);
+
+                        writeArkoudaMetaData(file_id, group, objType, getHDF5Type(real));
+                    } when (DType.Bool) {
+                        var valEntry:SymEntry = toSymEntry(toGenSymEntry(vals), bool);
+
+                        //localize values and write dataset
+                        var localVals = new lowLevelLocalizingSlice(valEntry.a, 0..#valEntry.size);
+                        var val_dims: C_HDF5.hsize_t = valEntry.size:C_HDF5.hsize_t;
+                        C_HDF5.H5LTmake_dataset(file_id, "/%s/values".format(group).c_str(), 1:c_int, val_dims, getDataType(bool), localVals.ptr);
+                        
+                        //localize segments and write dataset
+                        var localSegs = new lowLevelLocalizingSlice(segEntry.a, 0..#segEntry.size);
+                        var seg_dims: C_HDF5.hsize_t = segEntry.size:C_HDF5.hsize_t;
+                        C_HDF5.H5LTmake_dataset(file_id, "/%s/segments".format(group).c_str(), 1:c_int, seg_dims, getDataType(int), localSegs.ptr);
+
+                        writeArkoudaMetaData(file_id, group, objType, getDataType(bool));
+                        writeArkoudaMetaData(file_id, "%s/segments".format(group), "pdarray", getDataType(int));
+                        writeArkoudaMetaData(file_id, "%s/values".format(group), "pdarray", getDataType(bool));
+                    }
+                    otherwise {
+                        throw getErrorWithContext(
+                           msg="Unsupported SegArray DType %s".format(dtype2str(dType)),
+                           lineNumber=getLineNumber(),
+                           routineName=getRoutineName(), 
+                           moduleName=getModuleName(),
+                           errorClass="IllegalArgumentError");
+                    }
+                }
+                
+                C_HDF5.H5Fclose(file_id);
+            }
+            when MULTI_FILE {
```

Comment:
Again I think we can reduce the amount of code here with some structure changes to provide for easier maintenance in the future. 

I would also like to change the name of `writeStringsComponentToHdf` to `writeSegmentedComponentToHdf` so that it encompasses both scenarios. 

```chapel
// example of block within multifile
when MULTI_FILE {
    var filenames = prepFiles(filename, mode, segEntry.a);

    const lastSegIdx = segEntry.a.domain.high; // TODO remove

    select dType {
        when DType.Int64 {
            var valEntry:SymEntry = toSymEntry(toGenSymEntry(vals), int);
            ref sv = valEntry;
            const lastValIdx = valEntry.a.domain.high; // TODO - remove
            writeSegmentedDistDset(filenames, group, objType, valEntry.a, segEntry.a, st, int);
        }
        // ...
        otherwise {
            // error handling
        }
    }
}

// helper function definition
inline proc writeSegmentedDistDset(filenames: [] string, dset_name: string, objType: string, values, segments, st: borrowed SymTab, type t) throws {
    const lastSegIdx = segments.domain.high;
    const lastValIdx = values.domain.high;
    coforall (loc, idx) in zip(segments.targetLocales(), filenames.domain) do on loc {
        const localeFilename = filenames[idx];
        h5Logger.debug(getModuleName(),getRoutineName(),getLineNumber(),
                    "%s exists? %t".format(localeFilename, exists(localeFilename)));

        var file_id = C_HDF5.H5Fopen(localeFilename.c_str(), C_HDF5.H5F_ACC_RDWR, C_HDF5.H5P_DEFAULT);
        defer { // Close the file on scope exit
            C_HDF5.H5Fclose(file_id);
        }

        // create the group
        validateGroup(file_id, localeFilename, dset_name);

        // write the segments
        const locDom = segments.localSubdomain();
        var dims: [0..#1] C_HDF5.hsize_t;
        dims[0] = locDom.size: C_HDF5.hsize_t;

        var localSegs = segments[locDom];
        var startValIdx = localSegs[locDom.low];
        var endValIdx = if (lastSegIdx == locDom.high) then lastValIdx else A[locDom.high + 1] - 1;
        var valIdxRange = startValIdx..endValIdx;

        var localVals: [valIdxRange] int;

        ref olda = values.a;
        forall (localVal, valIdx) in zip(localVals, valIdxRange) with (var agg = newSrcAggregator(t)) {
            // Copy the remote value at index position valIdx to our local array
            agg.copy(localVal, olda[valIdx]); // in SrcAgg, the Right Hand Side is REMOTE
        }

        writeSegmentedComponentToHdf(file_id, group, SEGMENTED_VALUE_NAME, localVals);
        localSegs = localSegs - startValIdx;
        writeStringsComponentToHdf(file_id, group, SEGMENTED_OFFSET_NAME, localSegs);

        writeArkoudaMetaData(file_id, group, objType, getDataType(t));
    }
}

// i would update writeStringsComponentToHdf to the following definition
private proc writeSegmentedComponentToHdf(fileId: int, group: string, component: string, items: [] ?etype) throws {
    var numItems = items.size: uint(64);
    C_HDF5.H5LTmake_dataset_WAR(fileId, '/%s/%s'.format(group, component).c_str(), 1,
            c_ptrTo(numItems), getHDF5Type(etype), c_ptrTo(items));

    writeArkoudaMetaData(fileId, "%s/%s".format(group, component), "pdarray", getDataType(etype));
}
```

Reply:
Added this recommendation with tweaks for fixes found during debugging